### ISSUE TITLE
Move Enum mappings from @Mapping to an @ValueMapping 

### ISSUE THREAD

**Issue by sjaakd** [mapstruct/mapstruct#749]
I'm often in the situation that I need to map
String -> String
Integer -> Sting
String -> Integer
etc. 

The current way is to specify a qualifer. So you and up with 10 to 20 qualifiers for 20 String to String methods + a method implementation for each one of them. In other words, more code then you write it by hand. Qualifiers / methods are simply to clunky.

Proposal:
Make it possilbe to directly refrence a (immutable) Map to be used for such mappings.

**Comment by gunnarmorling** [comment]
Why is it that you need to specify a qualifier to begin with? Can you share
an example?

**Comment by sjaakd** [comment]
Well, I've got several methods that require a mapping from String to String. Each making use of its own (static) `Map`. To distinguish between them, I now need a qualifier to select the proper mapping.

Example coming up :smile:

**Comment by sjaakd** [comment]
Example:

```java

public class SourceClass {

    private String sourceMember;

    public String getSourceMember() {
        return sourceMember;
    }

    public void setSourceMember(String sourceMember) {
        this.sourceMember = sourceMember;
    }
   
}
```

``` java

public class TargetClass {

    private String targetMember;
    private String secondMember;
    
    public String getTargetMember() {
        return targetMember;
    }

    public void setTargetMember(String targetMember) {
        this.targetMember = targetMember;
    }

    public String getSecondMember() {
        return secondMember;
    }

    public void setSecondMember(String secondMember) {
        this.secondMember = secondMember;
    }
    
    
}

```

``` java
import java.util.HashMap;
import java.util.Map;

public class Source2TargetMapper {

    public static Map<String, String> secondMemberMapping;
    static {
        secondMemberMapping = new HashMap<String, String>();
        secondMemberMapping.put( "first", "1st" );
        secondMemberMapping.put( "second", "2nd" );
    }
    
    public TargetClass source2Target( SourceClass source ) {
        
        TargetClass target = new TargetClass();
        // Straight forward mapping: (could be done with @Mapping( target = "targetMember", source = "sourceMember" ) using mapstruct)
        target.setTargetMember( source.getSourceMember() );
        
        // can be mapped using mapstruct 'expression': 
        // @Mapping( target = "secondMember", expression = "java(Source2TargetMapper.secondMemberMapping.get(source.sourceMember))")
        // or by using a qualifier
        target.setSecondMember( secondMemberMapping.get( source.getSourceMember() ) );
        
        return target;
    }
}

```

``` java
public class Source2TargetMain {

    public static void main(String[] args) {
        
        SourceClass source = new SourceClass();
        source.setSourceMember( "second" );
        
        Source2TargetMapper mapper = new Source2TargetMapper();
        TargetClass target = mapper.source2Target( source );
        
        System.out.println( target.getTargetMember() ); // output: 'second'
        System.out.println( target.getSecondMember() ); // output: '2nd'
    }

}

```

**Comment by gunnarmorling** [comment]
How about this:

    @Qualfier
    public @interface CustomMapped {}


    public class StringMapper {

        private static Map<String, String> stringMappings;

        static { ... initialize ... }

        @CustomMapped
        String customMappedString(String source) {
            return stringMappings.get( source );
        }
    }

    @Mapper
    public interface Source2TargetMapper {

         @Mapping(target="secondMember", source="sourceMember", qualifiedBy=CustomMapped.class)
         public TargetClass source2Target( SourceClass source );
    }

I.e. you'd have a single qualifier which delegates mapping of that property to the map in a hand-written mapper class. That should do the trick?

**Comment by sjaakd** [comment]
> I.e. you'd have a single qualifier which delegates mapping of that property to the map in a hand-written mapper class. That should do the trick?

But that's what I meant with clunky.. I got not 1 but 20 qualifiers, each only used for one mapping.. This way I write far more code then I would by doing it by hand (assuming you also specify retention, target). I reckon this is a common problem.

**Comment by gunnarmorling** [comment]
That's the part I don't understand, why is it 20 qualifiers? It's just a
single one, @CustomMapped, which causes conversion of all the properties
qualified with it to be mapped via that custom method based on the map.

**Comment by sjaakd** [comment]
We've got 20 of these tables.. :smile:

**Comment by gunnarmorling** [comment]
Ah, now I can make better sense out of it. So why is it that you don't have
a single mapping table? Do you actually need to map the string "foo" to
"bar" in one use case but to "baz" in another?

**Comment by sjaakd** [comment]
> Do you actually need to map the string "foo" to "bar" in one use case but to "baz" in another?
More or less. I need mappings from "foo2" to "baz" . So the KVP's are entirely different for each mapping.

**Comment by gunnarmorling** [comment]
Do you have ideas how it should look API-wise? I am no huge fan of introducing the concept of maps, but we might find a way to simplify the selection of mapping methods in case several candidates qualify. 

E.g. with a `@Named("somemapping")` annotation and adding a `qualifiedByName("somemapping")` attribute to `@Mapping`.

**Comment by sjaakd** [comment]
Heavily inspired by CDI :smile: but that could help. 

I still believe that a `Map` is '_the  vehicle_' in Java to perform constant mappings from one 'type constant' to another 'type constant'. 

So, what if we would allow this `Named` annotation on both mapping methods (as we do now for qualifiers), but make MapStruct a bit smarter when it comes to maps. E..g also allow this annotation on constant maps (static final maps in an abstract class and maps in an interface) and let MapStruct do set the target to  `get( source )` on that map. That would also remove the need of writing a mapping method that does that get call.

**Comment by sjaakd** [comment]
@gunnarmorling: shall I split this issue in two separate issues?
1. Introduction of `@Named`, which  is a good idea on its own merrits..
2. Automatically calling a contant mapping by means of a (static final) `Map` in the `@Mapper` annotated class, which deserves some more discussion. (E.g. we could construct forged method).

**Comment by gunnarmorling** [comment]
I object to the direct usage of map for two reasons:

* it leads us into territories of accessing state rather than functionality and all the implications this brings with regard to thread-safety; Is the map safe for concurrent access, is it truely immutable, is it safely published to other threads? There is just too many assumptions we'd have to make; All this is much better addressed in a method of the user, where they can ensure the right level of synchronization etc.
* it's not necessarily the fastest you can do; For small numbers (I can't give an exact threshold, but think of < 10 or so), an if/then/elseif... cascade is quicker than doing hash map look-ups

In other words, using a map for this functionality is an "implementation detail" which MapStruct should not make any assumptions about. Going through a method is a bit more verbose but it's just the safer alternative.

**Comment by sjaakd** [comment]
> I object to the direct usage of map for two reasons:

Ok.. I see your points..

:bulb: What if we do something similar as we do for enum mappings.. So.. Introduce a new `@ConstantMappings` and entries `@ConstantMapping` along the line of what we have for enum mappings right now. Then we can generate the method body ourselves and take your objections into considerations. E.g. use if/then/elseif construction for < 10 mappings and guave immutable maps for >= 10 mappings.. 

In the future (2.0) the enum mapping can migrate to the same structure (and not be part of regular `@Mapping` anymore),  in other words, be a special case of the `@ConstantMappings` ...

**Comment by gunnarmorling** [comment]
That's an interesting idea. -1 for using Guava in generated code, though ;)

In hindsight, using `@Mapping` for enums was a mistake, a separate annotation such as `@ValueMapping` would have been the better alternative.

**Comment by sjaakd** [comment]
> -1 for using Guava in generated code, though ;)

Ahh.. I knew you were going to say that.. Guess my 'e.g.' was not careful enough :smile:

Still, I would like to go ahead with your `@Named` idea first (1) and then continue with exploring the idea of`@ConstantMappings` further.  Coming back on my previous points. Shall I open 2 follow up issues and refer to this one (and close this one).

**Comment by gunnarmorling** [comment]
> Still, I would like to go ahead with your @Named idea first (1)

+1 That should improve things for you already. Thanks for taking it up!

> Shall I open 2 follow up issues and refer to this one

As you prefer :) I'd keep this one for @ValueMapping for future reference
and open a new one for @Named.

**Comment by sjaakd** [comment]
> As you prefer :) I'd keep this one for @ValueMapping for future reference
> and open a new one for @Named.

Ah... and `@ValueMapping` is indeed a better name.. :+1:  

> +1 That should improve things for you already. Thanks for taking it up!

No problem.. It does.

**Comment by gunnarmorling** [comment]
Btw. for `@ValueMapping` (and I think it's the case for the current `constant()` mapping), we should invoke the String->XYZ conversion within MapStruct at compile time and not in the generated code. E.g. if a property is of type `int`, the generated code for referring to the constant or map entry for it should be `42` and not `Integer.valueOf( "42")`.

**Comment by sjaakd** [comment]
> tw. for @ValueMapping (and I think it's the case for the current constant() mapping), we should invoke the String->XYZ conversion within MapStruct at compile time and not in the generated code.

I realize that. We need to think how we handle enums in that perspective.

**Comment by sjaakd** [comment]
@gunnarmorling, @agudian : proposal for the API as specified below to solve value mappings in a more general way:


```java
@Retention(RetentionPolicy.CLASS)
@Target(ElementType.METHOD)
public @interface ValueMappings {

    ValueType sourceValueType() default ValueType.STRING;
    ValueType targetValueType() default ValueType.STRING;
    String defaultValue();
    ValueMapping[] value();

}

@Retention(RetentionPolicy.CLASS)
@Target(ElementType.METHOD)
public @interface ValueMapping {

    String target();
    String source();
}

public enum ValueType {

    BYTE,
    SHORT
    INT,
    LONG,
    FLOAT,
    DOUBLE,
    CHAR,
    STRING,
    BOOLEAN, /*?*/
    WRAPPED_BYTE,
    WRAPPED_SHORT
    WRAPPED_INT,
    WRAPPED_LONG,
    WRAPPED_FLOAT,
    WRAPPED_DOUBLE,
    WRAPPED_CHAR,
    WRAPPED_BOOLEAN, /*?*/
    BIG_INTEGER,
    BIG_DECIMAL,
    ENUM
}

/**
 * Configures the mapping of one bean attribute or enum constant.
 * <p>
 * The name of the mapped attribute or constant is to be specified via {@link #target()}. For mapped bean attributes it
 * is assumed by default that the attribute has the same name in the source bean. Alternatively, one of
 * {@link #source()}, {@link #expression()} or {@link #constant()} can be specified to define the property source.
 * <p>
 * In addition, the attributes {@link #dateFormat()} and {@link #qualifiedBy()} may be used to further define the
 * mapping.
* <p>
 * <b>IMPORTANT NOTE: mapping of enum constants will be deprecated from version 2.0, and replaced by  {@link ValueMapping}.</b> 
 *
 * @author Gunnar Morling
 */
@Retention(RetentionPolicy.CLASS)
@Target(ElementType.METHOD)
public @interface Mapping {
...
}
```

WDYT? I want to start working on this soon. I reckon that I can reuse some of the enum mapping stuff.

**Comment by gunnarmorling** [comment]
Hey @sjaakd, can you show an example of how the API would be used? I find myself often starting by writing down the code I'd like to write as an API user to achieve a given thing, and then I go and create the types/methods I "used" that way. Thanks!

**Comment by sjaakd** [comment]
Of course:

Example for String to String variant

```java
   @ValueMappings(  {
   @ValueMapping( source = "FOO", target = "BAR" ),
   @ValueMapping( source = "FOO1", target = "BAR1" )
} ) 
```
Example for integer to String variant:

```java
@ValueMappings( sourceValueType = ValueType.INTEGER, default="BAZ", values = {
   @ValueMapping( source = "1", target = "BAR" ),
   @ValueMapping( source = "2", target = "FOO" )
} ) 
```

enum to enum

```java
@ ValueMappings(  sourceValueType = ValueType.ENUM, targetTypeValue = ValueType.ENUM, values = {
    @ ValueMapping(source = "EXTRA", target = "SPECIAL"),
    @ ValueMapping(source = "STANDARD", target = "DEFAULT"),
    @ ValueMapping(source = "NORMAL", target = "DEFAULT")
} )
```

BTW: for implementation: the general idea is to generate a switch / case for all primitive source types and a cascading if-then-else for other types. I doubt about the usefulness of a boolean mapping, but perhaps for completeness reasons.

---------------

We could even get rid of the `@ValueMapping` perhaps, now I think of it a bit more, like this (perhaps a bit more rigid but less verbose).

```java
@Retention(RetentionPolicy.CLASS)
@Target(ElementType.METHOD)
public @interface ValueMappings {

    ValueType sourceValueType() default ValueType.STRING;
    ValueType targetValueType() default ValueType.STRING;
    String defaultValue();
    String[][]  mappings();

}
```

**Comment by gunnarmorling** [comment]
What is the rationale for sourceValueType and targetValueType()? Couldn't
these be derived from the parameter/return type of the annotated mapping
method?

**Comment by sjaakd** [comment]
> What is the rationale for sourceValueType and targetValueType()? Couldn't
> these be derived from the parameter/return type of the annotated mapping
> method?

You're absolutely right.. I just realized that myself as well :unamused: ..

**Comment by sjaakd** [comment]
@gunnarmorling  I think I've nailed it.

```java
/**
 * Constructs a set of value (constant) mappings.
 *
 * Each mapping entry is provided by a mapping in {@link #value()}. There are three scenarios to consider:
 * <ol>
 * <li>There is a non null source and there's a match</li>.
 * <li>The source provided is null</li>.
 * <li>There is no match</li>.
 *
 * {@link ValueMappingType} will be used to distinguish between those scenarios. The latter 2 scenarios may only appear
 * once in the provided values {@link #value() }
 * </ol>
 *
 * @author Sjaak Derksen
 */
@Target(ElementType.METHOD)
@Retention(RetentionPolicy.CLASS)
public @interface ValueMappings {

    ValueMapping[] value();

}
```

```java
/**
 * Configures the mapping of one constant value to another.
 * <p>
 * The the source constant is to be specified {@link #source()}. It is mapped to the provided {@link #target()}
 *
 * @author Sjaak Derksen
 */
@Repeatable(ValueMappings.class)
@Retention(RetentionPolicy.CLASS)
@Target(ElementType.METHOD)
public @interface ValueMapping {

    /**
     * Controls the type of mapping.
     *
     * Note: It is not allowed to define {@link #source()} for {@link ValueMappingType#DEFAULT} or
     * {@link ValueMappingType#NULL}
     * @return The mapping type.
     */
    ValueMappingType valueMappingType() default ValueMappingType.SOURCE;


    /**
     * The source value constant to use for this mapping.
     *
     * If the value is not of type {@code String}, the value will be converted by applying a matching
     * conversion method or built-in conversion.
     *
     * @return The source value.
     */
    String source() default "";

    /**
     * The target value constant to use for this mapping.
     *
     * If the value is not of type {@code String}, the value will be converted by applying a matching
     * conversion method or built-in conversion.
     *
     * @return The target value.
     */
    String target();

}
```
```java

/**
 * Type of value mapping.
 *
 * @author Sjaak Derksen
 */
public enum ValueMappingType {

    /**
     * The {@link #ValueMapping#source} will be mapped to {@link #ValueMapping#target} which will be returned.
     */
    SOURCE,

    /**
     * If no mapping is defined for the value passed to the mapping method. The {@link #ValueMapping#target} represents
     * the default value to return.
     */
    DEFAULT,

    /**
     * If {@code null} is passed to a mapping method, the {@link #ValueMapping#target} in this mapping will be returned
     */
    NULL;
}
```
Example for String to String variant

```java
@ValueMappings(  {
  @ValueMapping( valueMappingType = ValueMappingType.NULL, target="" ),
  @ValueMapping( source = "FOO", target = "BAR" ),
  @ValueMapping( source = "FOO", target = "BAR" ),
  @ValueMapping( source = "FOO1", target = "BAR1" )
  @ValueMapping( valueMappingType = = ValueMappingType.DEFAULT, target = "BAZ" )
} ) 
```

**Comment by sjaakd** [comment]
Of course: the javadoc needs some further refinement...

**Comment by gunnarmorling** [comment]
Thanks for the proposal.

Tbh. I am not convinced of the feature's usefulness. What is the advantage of "programming" this in annotations instead of writing an actual method? Granted, you save some characters, but you also loose things such as:

* Being able to refer to constants
* Direct type checking by the compiler; I think feedback when writing something like `int i = "String"` is better than when doing the same mistake via annotations where MapStruct needs to report this
* Being able to refer to expressions (e.g. String concatenations, method invocations, or e.g. a range of numbers you'd like to map to one String representing that range etc.)

For enums the mapping of values made more sense as one is dealing with a limited set of possible values on both sides, so MapStruct's approach of defaulting everything not explicitly mapped can be leveraged. But for "free value" types such as String or numbers you need to list every mapping in the annotations, which makes this much less attractive to me.

All in all I am having doubts that this feature pulls its weight at this point.

**Comment by sjaakd** [comment]
Ok. Points taken.

So lets limit ourself to two things initially:
* enums. I think they have to move away from the current mapping as we discussed earlier.
* Strings. They do not need a type conversion (and its the case I use most). 

Advantages:
* Reverse mappings come out of the box.
* The solutions objective is to add null mappings and default mappings to enum mappings. The complete on name-match feature  still will be in place.
* Value mappings are handled in a -clearly- distinct way from bean mappings, map mappings and iterable mappings.
* We can generate cascading if-then else or case statements (they require coding and breaks are easy forgotten). As you pointed out those are faster.
* You can handle your mapping-logic on one place (also when its an interface).
* We have a placeholder for value mappings if more requests are made. I do believe (from practical situations) that these kind of mappings are part of the business of a mapper, since I've been writing them at my current job, but also a lot in the past. For instance normalization of databases is carried out by data-base designers. DTO by a software designer. So especially String constants do tend to look slightly different. 
* In addition to last point we do have a placeholder for adding name normalization rules in the future in order to do map values automatically between a DTO and entity layer (e.g. I've cases were a name of an string constant ends with NM in de db, and not in the DTO, booleans are text values in the database (J or N, in dutch, and end with BLN, etc.) and boolean in the business DTO. I reckon I'm not the only one. I even think (don't pin me to it) I saw an issue for that popping up.

I would say, selecting only Strings and Enums would overcome your first two points.

WDYT?

**Comment by gunnarmorling** [comment]
Hi @sjaakd, some remarks:

> enums. I think they have to move away from the current mapping as we discussed earlier.

+1. We should have a dedicated `@EnumMapping(s)` so let's do that and deprecate the enum mapping support via `@Mapping`.

Regarding Strings I'm having some doubts still.

> Reverse mappings come out of the box.

That's cool for enums. For Strings, you also could use a Guava [BiMap](http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/collect/BiMap.html) in your custom implementation if it's a fully 1:1 bi-directional mapping.

> The solutions objective is to add null mappings...

Did you see PR #509? This should deal with that already. Can you check out on it to make sure it's considered properly? Maybe we also should just fix this in the new `@EnumMapping`, giving another incentive to not further use `@Mapping` for enum mappings.

> ...and default mappings to enum mappings.

We already have that, right? You'd just expose it via a new annotation I assume?

> For instance normalization of databases is carried out by data-base designers. DTO by a software designer. So especially String constants do tend to look slightly different. 

Yes, I absolutely understand the need for this sort of value mappings in applications. It's only that I don't think MapStruct would be of great help here with an annotation based approach. Value mappings (with the exception of enum mappings where defaulting makes sense often) seem just be much better addressed by custom code. I.e. let people deal with proper `Map`s or if/then/else themselves instead of letting them define "maps through annotations". Using annotations is great for stuff which is statically defined (property names, but also enum constants) and thus can be validated at compile time, but for values I don't see any advantage really.

> In addition to last point we do have a placeholder for adding name normalization rules in the future

I don't buy into the placeholder argument really :) Let's solve the issue when it really shows up. Then we know what's needed and can build a suitable solution for it. My experiences with building stuff in advance without clear need are not good ones.

All in all, I'd like us to support enum mappings via their own annotation (including null support) and deprecate enum mappings via `@Mapping`. I.e. a warning should be raised in this case. All other value mappings I'd leave to custom code.

**Comment by gunnarmorling** [comment]
+1 for naming it `@ValueMapping` in order to leave the door open for later generalizations. For now, only enum mappings would be supported.

One thing I was contemplating was to to provide support for more strongly typed enum mappings where one would use the actual enum constants in a custom annotation.

Say I have this:

    enum Source { FOO, BAR; }
    enum Target { BAZ, QUX; }

Then there could be a custom (user-defined, or we somehow generate it?) enum mapping annotation:

    @ValueMapping // marks this as custom enum mapping annotation
    @interface MyEnumMapping {
        @OverridesAnnotation("source") // marks this as representing @ValueMapping#source
        Source source();

        @OverridesAnnotation("targets") // marks this as representing @ValueMapping#target
        Target target();
    }

Then in mapping methods one could use the actual enum constants instead of Strings.

Not sure whether it's worth the effort, also handling nulls would be an issue. The override mechanism is similar to the one from Bean Validation (see "constraint composition in http://beanvalidation.org/1.1/spec/#constraintsdefinitionimplementation-constraintcomposition).

**Comment by sjaakd** [comment]
Its a pity that Annotations don't allow some kind of Generic mechanism. 

Anyway.. I'm not so sure mainly because of the complexity we introduce.
1. We detach the actually mapping definition from the spot were we have the mapping method.
2. Defining annotations is not the regular job of a programmer. It takes a learning curve.
3. If we are going to generate it, its not available initially (or after clean). You basically have to run generate-sources. Which adds complexity on its own.
4. How would a method declaration look like? 
5. Do we have to define the `List` (MyEnumMappings) as well?

Need to think of this a bit more as well.

**Comment by gunnarmorling** [comment]
Yes, it adds quite some complexity. Nothing we should do right now for
sure, I just liked the idea in a way :)

**Comment by sjaakd** [comment]
@gunnarmorling : I think I stay away from the `@ValueMapping` for the time being.. :smile:

Lets open up a new issue when we want to handle a specific `@ValueMapping` or.. are you convinced and shall I open up a follow up one for enum-string, string-enum and string-string ?

**Comment by gunnarmorling** [comment]
Your call; I'm not overly keen on adding these as you know, so staying away
from them would be my preference :)

**Comment by sjaakd** [comment]
So.. lets park this one for the time being.. I have some other stuff which is higher on my list as well for the time being.
